packages/dds/legacy-dds/src/sharedSignal.ts

```diff
@@ -0,0 +1,169 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+import type {
+	IChannelAttributes,
+	IFluidDataStoreRuntime,
+	IChannelStorageService,
+	IChannelFactory,
+} from "@fluidframework/datastore-definitions/internal";
+import { FileMode, MessageType, TreeEntry } from "@fluidframework/driver-definitions/internal";
+import type {
+	ISequencedDocumentMessage,
+	ITree,
+} from "@fluidframework/driver-definitions/internal";
+import type { ISummaryTreeWithStats } from "@fluidframework/runtime-definitions/internal";
+import { convertToSummaryTreeWithStats } from "@fluidframework/runtime-utils/internal";
+import type { IFluidSerializer } from "@fluidframework/shared-object-base/internal";
+import { SharedObject } from "@fluidframework/shared-object-base/internal";
+
+import type {
+	ISharedSignal,
+	ISharedSignalEvents,
+	ISignalOperation,
+	SerializableTypeForSharedSignal,
+} from "./interfaces.js";
+import { SharedSignalFactory } from "./sharedSignalFactory.js";
+
+const snapshotFileName = "header";
+
+/**
+ * Represents a shared signal that allows communication between distributed clients.
+ *
+ * @internal
+ */
+export class SharedSignal<T extends SerializableTypeForSharedSignal = any>
+	extends SharedObject<ISharedSignalEvents<T>>
+	implements ISharedSignal<T>
+{
+	/**
+	 * Create a new shared signal
+	 *
+	 * @param runtime - data store runtime the new shared signal belongs to
+	 * @param id - optional name of the shared signal
+	 * @returns newly create shared signal (but not attached yet)
+	 */
+	public static create(runtime: IFluidDataStoreRuntime, id?: string): SharedSignal {
+		return runtime.createChannel(id, SharedSignalFactory.Type) as SharedSignal;
+	}
+
+	/**
+	 * Get a factory for SharedSignal to register with the data store.
+	 *
+	 * @returns a factory that creates and load SharedSignal
+	 */
+	public static getFactory(): IChannelFactory {
+		return new SharedSignalFactory();
```

Comment:
Not for this PR - since the factory is stateless, would be good to avoid instantiating a new one each time getFactory() is called.